cloudbuild/python-scripts/parse_logs.py

```diff
@@ -0,0 +1,513 @@
+    # 'application_YYYYYY_XXXX'
+    # We need to extract the actual application number within the cluster,
+    # i.e. the ending digits which is the job number in the cluster.
+    # The pattern
+    #     [^_]+: Searches for the last occurrence of characters
+    #     which are not a "_".
+    #     i.e. the job number present after the last "_"
+    # Thus the current pattern helps in searching "XXXX" as the part of group 0.
+
+    yarn_application_number_pattern = r'[^_]+$'
+
+    yarn_job_number = re.search(
+        yarn_application_number_pattern, yarn_application_number
+    ).group(0)
+
+    logs_pattern = (
+        f'{cluster_id}/yarn-logs/root'
+        f'/bucket-logs-ifile/{yarn_job_number}/{yarn_application_number}'
+    )
+    return logs_pattern
+
+
+def get_cluster_temp_bucket(client_project_name, cluster_name, region):
+    cluster_client = dataproc_v1.ClusterControllerClient(
+        client_options={'api_endpoint': f'{region}-dataproc.googleapis.com:443'}
+    )
+    get_cluster_request = dataproc_v1.GetClusterRequest(
+        project_id=client_project_name, region=region, cluster_name=cluster_name
+    )
+    cluster_resource = cluster_client.get_cluster(get_cluster_request)
+    return cluster_resource.config.temp_bucket
+
+
+def get_bucket_contents(bucket):
+    storage_client = storage.Client()
+
+    # Note: Client.list_blobs requires at least package version 1.17.0.
+    blobs = storage_client.list_blobs(bucket)
+    bucket_contents = []
+    # Note: The call returns a response only when the iterator is consumed.
+    for blob in blobs:
+        bucket_contents.append(blob.name)
+    return bucket_contents
+
+
+def read_logs(cluster_temp_bucket, logs_pattern, query):
+    """Method to parse the number of records from the yarn logs.
+
+    Args:
+        cluster_temp_bucket: name of the temp bucket that the dataproc cluster
+          uses to store yarn logs.
+        logs_pattern: pattern in which logs are found inside the cluster temp
+          bucket.
+        query: query set by the user to be executed before the BQ table read.
+
+    Returns:
+        Sum of metric values in all the files.
+
+    Raises:
+        RuntimeError: In case metric value is not found in any of the files.
+        Incase query has been set and query results were not found in any of the
+        log files.
+    """
+
+    metric_string = 'Number of records read: '
+    end_of_metric_string = ';'
+
+    # Sum across all the worker files.
+    total_metric_count = 0
+    # Check if metric is present in at least one of the files.
+    is_metric_found = False
+    is_query_result_found = False
+    # Get all the contents in the GCS Bucket.
+    gcs_bucket_contents = get_bucket_contents(cluster_temp_bucket)
+    # Form the logs_pattern
+
+    # logs are stored in files having names of the format 'log_pattern/...'
+    # the pattern enables searching for the same.
+
+    logs_pattern = re.compile(rf'{re.escape(logs_pattern)}/.+')
```

Comment:
Let's avoid unnecessary blank lines in comments and between comment and code

Reply:
Fixed.